### PR TITLE
Use Pip 21.* to install airflow officially

### PR DESCRIPTION
**Pull request by potiuk**
The PIP 20.2.4 was so far the only officially supported installation
mechanism for Airflow as there were some problems with conflicting
dependencies (which were ignored by previous versio of PIP).

This change attempts to solve this by removing a [gcp] extra
from `apache-beam` which turns out to be the major source of
the problem - as it contains requirements to the old version of
google client libraries (but apparently only used for tests).

The "apache-beam" provider migh however need the [gcp] extra
for other components so in order to not break the backwards
compatibility, another approach is used.

Instead of adding [gcp] as extra in the apache-beam extra,
the apache.beam provider's [google] extra is extended with
'apache-beam[gcp]' additional requirement so that whenever the
provider is installed, the apache-beam with [gcp] extra is installed
as well.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Comment by potiuk** [comment]
Hey @ashb and @kaxil -> If we get this one merged, I think we might want to release a few providers (Google, Azure) where we had some small issues with warning (I am looking at an Azure warning in a moment) + Apache.Beam operator with the updated extra and we might aim to release 2.0.3 rather quickly - with mostly "pip 21 full support".

**Comment by potiuk** [comment]
It's green. Looks like, when we merge, we have a full PIP 21 support for Airflow @kaxil @ashb

File: setup.py [hunk 1]
```diff
@@ -198,7 +198,7 @@ def get_sphinx_theme_version() -> str:
     'watchtower~=0.7.3',
 ]
 apache_beam = [
-    'apache-beam[gcp]',
+    'apache-beam>=2.20.0',
```
Comment (mik-laj) [inline setup.py:201]: gcp extra is required by Dataflow operators.

Comment (potiuk) [reply setup.py:201]: Good point. I added it to additional-extras for "apache.beam" in google provider. As the result this is what google provider extras will look like:

```
        extras_require={
            'amazon': ['apache-airflow-providers-amazon'],
            'apache.beam': ['apache-airflow-providers-apache-beam', 'apache-beam[gcp]'],
            'apache.cassandra': ['apache-airflow-providers-apache-cassandra'],
            'cncf.kubernetes': ['apache-airflow-providers-cncf-kubernetes'],
            'facebook': ['apache-airflow-providers-facebook'],
            'microsoft.azure': ['apache-airflow-providers-microsoft-azure'],
            'microsoft.mssql': ['apache-airflow-providers-microsoft-mssql'],
            'mysql': ['apache-airflow-providers-mysql'],
            'oracle': ['apache-airflow-providers-oracle'],
            'postgres': ['apache-airflow-providers-postgres'],
            'presto': ['apache-airflow-providers-presto'],
            'salesforce': ['apache-airflow-providers-salesforce'],
            'sftp': ['apache-airflow-providers-sftp'],
            'ssh': ['apache-airflow-providers-ssh'],
            'trino': ['apache-airflow-providers-trino'],
        },
```

Anyone willing to use DataFlow operators will have to use `apache-airflow-providers-google[apache.beam]`. Those extras are nicely documented in the README,

File: airflow/providers/apache/beam/CHANGELOG.rst [hunk 2]
```diff
@@ -19,6 +19,64 @@
...
+
+.. code-block:: python
+
+        extras_require={'google': ['apache-airflow-providers-google', 'apache-beam[gcp]']},
+
+You can still run this with PIP version <= 20.2.4 and go back to the previous behaviour:
+
+.. code-block:: shell
+
+  pip install apache-airflow-providers-google['apache.beam']
```
Comment (uranusjr) [inline airflow/providers/apache/beam/CHANGELOG.rst]:
```suggestion
  pip install apache-airflow-providers-google[apache.beam]
```

File: airflow/providers/apache/beam/CHANGELOG.rst [hunk 3]
```diff
@@ -19,6 +19,64 @@
...
+
+or
+
+.. code-block:: shell
+
+  pip install apache-airflow-providers-apache-beam['google']
```
Comment (uranusjr) [inline airflow/providers/apache/beam/CHANGELOG.rst]:
```suggestion
  pip install apache-airflow-providers-apache-beam[google]
```

File: airflow/providers/google/CHANGELOG.rst [hunk 4]
```diff
@@ -19,6 +19,64 @@
...
+
+.. code-block:: python
+
+        extras_require={'google': ['apache-airflow-providers-google', 'apache-beam[gcp]']},
+
+You can still run this with PIP version <= 20.2.4 and go back to the previous behaviour:
+
+.. code-block:: shell
+
+  pip install apache-airflow-providers-google['apache.beam']
```
Comment (uranusjr) [inline airflow/providers/google/CHANGELOG.rst]:
```suggestion
  pip install apache-airflow-providers-google[apache.beam]
```

File: airflow/providers/google/CHANGELOG.rst [hunk 5]
```diff
@@ -19,6 +19,64 @@
...
+
+or
+
+.. code-block:: shell
+
+  pip install apache-airflow-providers-apache-beam['google']
```
Comment (uranusjr) [inline airflow/providers/google/CHANGELOG.rst]:
```suggestion
  pip install apache-airflow-providers-apache-beam[google]
```

File: setup.py [hunk 6]
```diff
@@ -502,7 +502,7 @@ def get_sphinx_theme_version() -> str:
     'paramiko',
     'pipdeptree',
     'pre-commit',
-    'pylint>=2.7.0',
+    'pylint~=2.7.4',
```
Comment (uranusjr) [inline setup.py:505]: Is this a required change or drive-by refactoring?

Comment (potiuk) [reply setup.py:505]: If we don't do it, we will need to fix new issues reported by Pylint 2.8.0 released yesterday. 


General context: 

All the reqular PR's are using the current 'constraints'. But whenever we modify setup.py, such PR will automatically attempt to run "eager upgrade" to get latest versions of everything (including pylint new version in this case). This is our way to - pretty much automatically - upgrade all such constraints and detect any future incompatibilies early (but without affecting the regular PRs).

Context for pylint: 

We've already had similar problem caused by new pylint versions (new errors reported on already 'good' code whenever setup.py changedf) so I'd rather fix it now to minor version and upgrade it separately in a separate PR (especially that they are preparing to release a major 3.0 upgrade which might require a bit more 'overhaul'.

File: airflow/providers/apache/beam/CHANGELOG.rst (hunk 2 above)
Comment (potiuk) [reply airflow/providers/apache/beam/CHANGELOG.rst]: Right :)

**Comment by potiuk** [comment]
I'd love that one to be merged, to finally get PIP 21 support (and keep the deps under control).

**Comment by potiuk** [comment]
I'd love that one to be merged, to finally get PIP 21 support (and keep the deps under control).

**Review by kaxil — APPROVED** [review]

**Review by uranusjr — APPROVED** [review]

**Comment by uranusjr** [comment]
Just a heads-up, pip 21.1.1 will likely be released next week to fix an import issue for 3.6.0 and 3.6.1.

**Comment by potiuk** [comment]
> Just a heads-up, pip 21.1.1 will likely be released next week to fix an import issue for 3.6.0 and 3.6.1.

Yep. I think I will add a couple of "automated" warnings (a'la dependabot). that will make us aware of the need to upgrade when newer versions of PIP/Pylint and few other tools that we are pinning. Dependabot is not good for us, because we are continuously updating automatically, and keep open dependencies, but those few dependencies we have pinned might make perfect sense to get "warnings" about.
